File: clients/ios/TdgLoader/Pack.swift
import Foundation

/// The manifest is the contract with the generator. The app parses it and
/// nothing else: every item already carries the URL to fetch it from, so the
/// loader never needs to know how the server lays out paths.
struct PackItem: Identifiable {
    let name: String
    let kind: String            // "image" | "video"
    let bytes: Int64
    let sha256: String
    let takenAt: Date
    let url: URL
    let latitude: Double?
    let longitude: Double?

    var id: String { name }
    var isVideo: Bool { kind == "video" }
}

struct Pack {
    let jobId: String
    let album: String
    let fileCount: Int
    let totalBytes: Int64
    let items: [PackItem]

    /// Schema v2 writes `taken_at_utc` with an explicit offset. v1 wrote a
    /// naive local time, which is the bug that made a pack land at a different
    /// instant on every machine that imported it — parse it, but only as a
    /// fallback, and treat it as local time the way Photos itself would.
    private static func parseDate(_ s: String) -> Date {
        let iso = ISO8601DateFormatter()
        iso.formatOptions = [.withInternetDateTime]
        if let d = iso.date(from: s) { return d }
        iso.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        if let d = iso.date(from: s) { return d }
        let naive = DateFormatter()
        naive.dateFormat = "yyyy-MM-dd'T'HH:mm:ss"
        naive.timeZone = TimeZone.current
        return naive.date(from: s) ?? Date()
    }

    static func parse(_ data: Data, base: URL) throws -> Pack {
        guard let o = try JSONSerialization.jsonObject(with: data) as? [String: Any],
              let rawItems = o["items"] as? [[String: Any]] else {
            throw LoaderError.badManifest
        }
        let jobId = o["job_id"] as? String ?? "unknown"
        var items: [PackItem] = []
        items.reserveCapacity(rawItems.count)
        for it in rawItems {
            guard let name = it["name"] as? String,
                  let rel = it["url"] as? String,
                  let url = URL(string: rel, relativeTo: base)?.absoluteURL else { continue }
            let gps = it["gps"] as? [String: Any]
            items.append(PackItem(
                name: name,
                kind: it["kind"] as? String ?? "image",
                bytes: (it["bytes"] as? NSNumber)?.int64Value ?? 0,
                sha256: it["sha256"] as? String ?? "",
                takenAt: parseDate((it["taken_at_utc"] as? String)
                                   ?? (it["taken_at"] as? String) ?? ""),
                url: url,
                latitude: gps?["lat"] as? Double,
                longitude: gps?["lon"] as? Double))
        }
        return Pack(
            jobId: jobId,
            album: o["album"] as? String ?? "TDG \(jobId)",
            fileCount: o["file_count"] as? Int ?? items.count,
            totalBytes: (o["total_bytes"] as? NSNumber)?.int64Value
                        ?? items.reduce(0) { $0 + $1.bytes },
            items: items)
    }
}

/// What the six-digit pairing code resolves to.
struct Pairing {
    let jobId: String
    let label: String
    let fileCount: Int
    let totalBytes: Int64
    let manifestURL: URL

    static func parse(_ data: Data, base: URL) throws -> Pairing {
        guard let o = try JSONSerialization.jsonObject(with: data) as? [String: Any],
              let rel = o["manifest_url"] as? String,
              let url = URL(string: rel, relativeTo: base)?.absoluteURL else {
            throw LoaderError.badPairing
        }
        return Pairing(
            jobId: o["job_id"] as? String ?? "?",
            label: o["label"] as? String ?? "",
            fileCount: o["file_count"] as? Int ?? 0,
            totalBytes: (o["total_bytes"] as? NSNumber)?.int64Value ?? 0,
            manifestURL: url)
    }
}

enum LoaderError: LocalizedError {
    case badManifest
    case badPairing
    case notAuthorised
    case notAuthorisedFull
    case http(Int, String)
    case shortRead(String, Int64, Int64)
    case noSpace(Int64, Int64)

    var errorDescription: String? {
        switch self {
        case .badManifest: return "That manifest is not in a shape this app understands."
        case .badPairing: return "No finished pack has that code."
        case .notAuthorised:
            return "Permission to add photos was declined."
        case .notAuthorisedFull:
            return "Removing assets needs Full Access to the photo library. "
                 + "Adding them only needs add-only permission, so iOS asks "
                 + "again here — allow it, or delete the TDG album by hand."
        case .http(let code, let url): return "HTTP \(code) from \(url)"
        case .shortRead(let name, let want, let got):
            return "\(name): expected \(want) bytes, got \(got)"
        case .noSpace(let need, let free):
            return "Not enough space: needs \(ByteFormat.human(need)), "
                 + "\(ByteFormat.human(free)) free"
        }
    }
}

enum ByteFormat {
    static func human(_ n: Int64) -> String {
        if n < 0 { return "unknown" }
        let units = ["B", "KB", "MB", "GB", "TB"]
        var v = Double(n), i = 0
        while v >= 1024 && i < units.count - 1 { v /= 1024; i += 1 }
        return i == 0 ? "\(n) B" : String(format: "%.1f %@", v, units[i])
    }
}

File: clients/android/app/src/main/kotlin/com/tdg/loader/Pack.kt
package com.tdg.loader

import org.json.JSONObject
import java.net.URL

/**
 * The manifest is the contract with the generator. The app parses it and
 * nothing else: every item already carries the URL to fetch it from, so the
 * loader never has to know how the server lays out paths.
 */
data class PackItem(
    val name: String,
    val kind: String,          // "image" | "video"
    val bytes: Long,
    val sha256: String,
    val takenAtUtc: String,    // ISO-8601, always with an offset since schema v2
    val url: String,
) {
    val isVideo get() = kind == "video"

    val mimeType: String
        get() = when {
            isVideo -> "video/mp4"
            name.endsWith(".heic", true) -> "image/heic"
            name.endsWith(".png", true) -> "image/png"
            else -> "image/jpeg"
        }

    /** Epoch millis, which is what MediaStore's DATE_TAKEN wants. */
    fun takenAtMillis(): Long =
        try {
            java.time.OffsetDateTime.parse(takenAtUtc).toInstant().toEpochMilli()
        } catch (e: Exception) {
            // A v1 manifest has no offset. Falling back to the local zone is
            // the same guess iOS Photos makes, and is why schema v2 exists.
            try {
                java.time.LocalDateTime.parse(takenAtUtc)
                    .atZone(java.time.ZoneId.systemDefault()).toInstant().toEpochMilli()
            } catch (e2: Exception) {
                System.currentTimeMillis()
            }
        }
}

data class Pack(
    val jobId: String,
    val album: String,
    val fileCount: Int,
    val totalBytes: Long,
    val items: List<PackItem>,
) {
    /** DCIM/<album> — a folder most galleries surface as an album. */
    val relativePath get() = "DCIM/$album"

    companion object {
        fun parse(json: String, base: URL): Pack {
            val o = JSONObject(json)
            val arr = o.getJSONArray("items")
            val items = ArrayList<PackItem>(arr.length())
            for (i in 0 until arr.length()) {
                val it = arr.getJSONObject(i)
                val rel = it.optString("url", "")
                items.add(
                    PackItem(
                        name = it.getString("name"),
                        kind = it.optString("kind", "image"),
                        bytes = it.getLong("bytes"),
                        sha256 = it.optString("sha256", ""),
                        takenAtUtc = it.optString("taken_at_utc", it.optString("taken_at", "")),
                        // Item URLs are server-relative; resolve against the
                        // manifest's own URL so the app carries no host config.
                        url = if (rel.startsWith("http")) rel else URL(base, rel).toString(),
                    )
                )
            }
            return Pack(
                jobId = o.getString("job_id"),
                album = o.optString("album", "TDG ${o.getString("job_id")}"),
                fileCount = o.optInt("file_count", items.size),
                totalBytes = o.optLong("total_bytes", items.sumOf { it.bytes }),
                items = items,
            )
        }
    }
}
